### PR TITLE
[JENKINS-50211] Reject connections from agents with unsupported Remoting versions

### PR DESCRIPTION
**Pull request by stellargo**
…upported Remoting versions

See [JENKINS-50211](https://issues.jenkins-ci.org/browse/JENKINS-50211).

Initial work : https://github.com/jenkinsci/jenkins/pull/3353.

Changes:

Incorporated reviews.
Fixed issue: The configuration was not persisting.
Fixed issue: The connection was not being closed.
Moved the code from Hudson to Jenkins.
Fixed the getter function of `RemoteConfiguration`.
Removed unnecessary functions.
Added documentation.

### Proposed changelog entries

* Reject connections from agents with unsupported Remoting versions.

### Proposed upgrade guidelines

Jenkins now rejects connections from agents with unsupported Remoting versions. This can be disabled by setting the `hudson.slaves.SlaveComputer.allowUnsupportedRemotingVersions` system property to true.

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/java/hudson/slaves/SlaveComputer.java
Comment (stellargo) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: This closes the channel, however the event for connecting agent to Jenkins keeps getting triggered. This leads to a sort of loop. It would be really helpful if somebody could advice me on how to break this loop.

**Comment by daniel-beck** [comment]
Why would this be an option in the first place? Just reject and done?

**Comment by jeffret-b** [comment]
@stellargo , the PR looks good and looks like it meets the acceptance criteria Oleg laid out in the Jira ticket.

But, Daniel raises a good point. I think the default should be to reject connections that don't meet the minimum requirement. We can add a property to override that behavior. When the version was first implemented, there was more reason to allow agents with unknown version. I don't think that exists any longer. If Oleg agrees with this approach, you could trim down this PR -- remove the UI pieces and add the property.

**Comment by daniel-beck** [comment]
> We can add a property to override that behavior.

Specifically https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties i.e. a hidden "escape hatch" rather than a supported option with UI.

**Comment by stellargo** [comment]
IIUC, that would make it non configurable via JCasC too. I just wanted to put this out there, before I make any changes, just to confirm.

**Comment by jeffret-b** [comment]
Thanks for pointing that out, @stellargo. I'm still of the opinion that it would be better to keep this as a hidden escape hatch rather than create a UI for it. There are many things in Jenkins for obscure use cases that use this approach. They are not configurable via JCasc either.

@oleg-nenashev, since you wrote the original acceptance criteria, what do you think about this alternative proposal?

**Comment by daniel-beck** [comment]
> IIUC, that would make it non configurable via JCasC too.

Not JCasC, but whatever config management defines the Jenkins invocation would take care of this. Depending on the exact implementation, might also be doable in a Groovy post launch script.

**Comment by oleg-nenashev** [comment]
Sorry, I am still catching up wit reviews

**Comment by alecharp** [comment]
@oleg-nenashev would you have some time to get back on this PR? thanks!

**Comment by varyvol** [comment]
@stellargo could you please resolve the merge conflict? Thanks!

**Comment by stellargo** [comment]
@varyvol done :)

File: core/src/main/java/hudson/slaves/SlaveComputer.java
Comment (varyvol) [inline core/src/main/java/hudson/slaves/SlaveComputer.java:144]: This is really a new possibility than a escape hatch, right? Because the default option is still the current one.

File: core/src/main/java/hudson/slaves/SlaveComputer.java
Comment (Wadeck) [inline core/src/main/java/hudson/slaves/SlaveComputer.java:147]:
```suggestion
    private static boolean REJECT_CONNECTION = SystemProperties.getBoolean(SlaveComputer.class.getName() + ".allowUnsupportedAgentConnection");
```

Otherwise you will be hit by @jtnord ;)

**Comment by jeffret-b** [comment]
It would be good to have a test or two here.

File: core/src/main/java/hudson/slaves/SlaveComputer.java
Comment (stellargo) [reply core/src/main/java/hudson/slaves/SlaveComputer.java:144]: Yeah, the default is still the current one. Would you prefer to have it renamed to something other than a escape hatch?

**Comment by jeffret-b** [comment]
I still think the default should go the other -- make it reject by default and make it an escape hatch to get the old behavior back. But, I won't block the change for that. It's still an improvement.

File: core/src/main/java/hudson/slaves/SlaveComputer.java
Comment (jeffret-b) [reply core/src/main/java/hudson/slaves/SlaveComputer.java:144]: As it is now, it's not an escape hatch, so it should be called something else -- something about activating the additional validation.

**Comment by alecharp** [comment]
There are conflicts to be taking care of. I also added the `proposed-for-close` as the last message on this PR was in october 2020.

**Comment by basil** [comment]
I merged, addressed review feedback, and added test coverage. This should be good to go now.

File: core/src/main/java/hudson/slaves/SlaveComputer.java
Comment (timja) [inline core/src/main/java/hudson/slaves/SlaveComputer.java:676]: The above reads a bit weird to be, 'will not' and then compatibility is NOT.
How about this instead?
```suggestion
                            + " The connection will be allowed, but compatibility is NOT"
```

**Review by timja — APPROVED** [review]

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by timja — APPROVED** [review]

**Comment by basil** [comment]
And a week after this got merged, [a user running a 2017 version of Remoting](https://issues.jenkins.io/browse/JENKINS-68271) got a failure and upgraded Remoting. Working as intended!

**Pull request merged**
